### PR TITLE
[feat] Add generic TorchvisionResnetImageEncoder in ImageEncoderFactory.

### PR DESCRIPTION
**Pull request by kdexd**
This PR adds a generic image encoder named `TorchvisionResnetImageEncoder` in `mmf.modules.ImageEncoderFactory` — this encoder can instantiate any ResNet-like model from [`torchvision.models`](https://pytorch.org/docs/stable/torchvision/models.html) based on its name. The encoder returns spatial grid features same as torchvision.

It can be instantiated using a minimal config as:

```yaml
type: "torchvision_resnet"
params:
    name: <name>
    pretrained: false
```

- Set `pretrained: true` for loading ImageNet-pretrained weights from torchvision.
- Supported `<name>`s are:
    - `resnet18`, `resnet34`, `resnet50`, `resnet101`, `resnet152`
    - `resnext50_32x4d`, `resnext101_32x8d`
    - `wide_resnet50_2`, `wide_resnet101_2`

### Example

```python
>>> import torch
>>> from omegaconf import OmegaConf
>>> from mmf.utils.build import build_encoder
>>> 
>>> config = OmegaConf.load("encoder_config.yaml")
>>> encoder = build_encoder(config)
>>> images = torch.randn(32, 3, 224, 224)   # shape: (b, c, h, w)
>>> grid_feats = encoder(images)  # shape: (b, 2048, 7, 7)
```

### Backward Compatibility

This PR maintains backward compatibility as it adds a new encoder without changing existing ones. However, it subsumes the functionality of [ResNet152ImageEncoder](https://github.com/facebookresearch/mmf/blob/d04914a7bbae28f33356e8b17113a81d6291a5dc/mmf/modules/encoders.py#L215), hence the latter could be later replaced and deprecated in favor of this addition.

### CODE REVIEW

**Comment by kdexd** [comment]
Additional Note: This encoder will not support other torchvision models like VGG, GoogleNet, ShuffleNet etc. — these classes have a different module naming convention and different output feature dimensions. Supporting all models in a single class may get too cluttered. We can introduce the functionality in a separate encoder for these model families if needed.

**Comment by vedanuj** [comment]
This looks good. Can you please rebase the PR?

**Comment by kdexd** [comment]
@vedanuj: Sounds good. I had one thought about the compatibility with the rest of MMF codebase. Do you think it may be useful to add a config parameter to control the format of output features — `"grid"` (BCHW, b x 2048 x 7 x 7), `"pooled"` (b x 2048) or `"vectors"` (b x 49 x 2048).

**Comment by vedanuj** [comment]
> @vedanuj: Sounds good. I had one thought about the compatibility with the rest of MMF codebase. Do you think it may be useful to add a config parameter to control the format of output features — `"grid"` (BCHW, b x 2048 x 7 x 7), `"pooled"` (b x 2048) or `"vectors"` (b x 49 x 2048).

I think this format is fine. I would leave it to the model to decide what it wants to do with the output from the encoder. I know few existing encoders handle the output format, but I feel that is not the right design. In the long run it will lead to  encoders getting bloated with model specific output format handling logic. Let me know if you think otherwise.

**Comment by kdexd** [comment]
Makes sense, then this PR is finished from my end!
